dvc/scm.py

```diff
@@ -123,3 +126,61 @@ def resolve_rev(scm: "Git", rev: str) -> str:
             if len(ref_infos) > 1:
                 raise RevError(f"ambiguous Git revision '{rev}'")
         raise RevError(str(exc))
+
+
+def get_revs_from_args(
+    scm: "Git",
+    head_revs: Optional[List[str]] = None,
+    num: int = 1,
+    all_branches: bool = False,
+    all_tags: bool = False,
+    all_commits: bool = False,
+) -> OrderedDict:
+    from dvc.repo.experiments.utils import fix_exp_head
+
+    if num < 1 and num != -1:
+        raise InvalidArgumentError(f"Invalid number of commits '{num}'")
+
+    if not any([head_revs, all_branches, all_tags, all_commits]):
+        return OrderedDict()
+
+    head_revs = head_revs or []
+    revs = []
+    for rev in head_revs:
+        n = 0
+        while True:
+            if num == n:
+                break
+            try:
+                head = fix_exp_head(scm, f"{rev}~{n}")
+                assert head
+                revs.append(resolve_rev(scm, head))
+            except RevError:
+                break
+            n += 1
```

Comment:
This `fix_exp_head` logic could be moved into the generalized `iter_rev_names` method (and `brancher()` by extension).

Everywhere we use `brancher` in DVC we also need the exp HEAD fixup. so we can do the `fix_exp_head` inside the refactored iterator and then remove some of the existing `fix_exp_head` calls in DVC (`exp show` + params/metrics/plots diff) so that it's always unified in one place